usr/src/lib/smbclnt/libfknsmb/common/fake_stream.c

```diff
@@ -0,0 +1,1834 @@
+#endif	/* _KERNEL */
+
+/*
+ * Patchable mblk/dblk kmem_cache flags.
+ */
+int dblk_kmem_flags = 0;
+int mblk_kmem_flags = 0;
+
+static int
+dblk_constructor(void *buf, void *cdrarg, int kmflags)
+{
+	dblk_t *dbp = buf;
+	ssize_t msg_size = (ssize_t)cdrarg;
+	size_t index;
+
+	ASSERT(msg_size != 0);
+
+	index = (msg_size - 1) >> DBLK_SIZE_SHIFT;
+
+	ASSERT(index < (DBLK_MAX_CACHE >> DBLK_SIZE_SHIFT));
+
+	if ((dbp->db_mblk = kmem_cache_alloc(mblk_cache, kmflags)) == NULL)
+		return (-1);
+	if ((msg_size & PAGEOFFSET) == 0) {
+		dbp->db_base = kmem_alloc(msg_size, kmflags);
+		if (dbp->db_base == NULL) {
+			kmem_cache_free(mblk_cache, dbp->db_mblk);
+			return (-1);
+		}
+	} else {
+		dbp->db_base = (unsigned char *)&dbp[1];
+	}
+
+	dbp->db_mblk->b_datap = dbp;
+	dbp->db_cache = dblk_cache[index];
+	dbp->db_lim = dbp->db_base + msg_size;
+	dbp->db_free = dbp->db_lastfree = dblk_lastfree;
+	dbp->db_frtnp = NULL;
+	dbp->db_fthdr = NULL;
+	dbp->db_credp = NULL;
+	dbp->db_cpid = -1;
+	dbp->db_struioflag = 0;
+	dbp->db_struioun.cksum.flags = 0;
+	return (0);
+}
+
+/*ARGSUSED*/
+static int
+dblk_esb_constructor(void *buf, void *cdrarg, int kmflags)
+{
+	dblk_t *dbp = buf;
+
+	if ((dbp->db_mblk = kmem_cache_alloc(mblk_cache, kmflags)) == NULL)
+		return (-1);
+	dbp->db_mblk->b_datap = dbp;
+	dbp->db_cache = dblk_esb_cache;
+	dbp->db_fthdr = NULL;
+	dbp->db_credp = NULL;
+	dbp->db_cpid = -1;
+	dbp->db_struioflag = 0;
+	dbp->db_struioun.cksum.flags = 0;
+	return (0);
+}
+
+static int
+bcache_dblk_constructor(void *buf, void *cdrarg, int kmflags)
+{
+	dblk_t *dbp = buf;
+	bcache_t *bcp = cdrarg;
+
+	if ((dbp->db_mblk = kmem_cache_alloc(mblk_cache, kmflags)) == NULL)
+		return (-1);
+
+	dbp->db_base = kmem_cache_alloc(bcp->buffer_cache, kmflags);
+	if (dbp->db_base == NULL) {
+		kmem_cache_free(mblk_cache, dbp->db_mblk);
+		return (-1);
+	}
+
+	dbp->db_mblk->b_datap = dbp;
+	dbp->db_cache = (void *)bcp;
+	dbp->db_lim = dbp->db_base + bcp->size;
+	dbp->db_free = dbp->db_lastfree = bcache_dblk_lastfree;
+	dbp->db_frtnp = NULL;
+	dbp->db_fthdr = NULL;
+	dbp->db_credp = NULL;
+	dbp->db_cpid = -1;
+	dbp->db_struioflag = 0;
+	dbp->db_struioun.cksum.flags = 0;
+	return (0);
+}
+
+/*ARGSUSED*/
+static void
+dblk_destructor(void *buf, void *cdrarg)
+{
+	dblk_t *dbp = buf;
+	ssize_t msg_size = (ssize_t)cdrarg;
+
+	ASSERT(dbp->db_mblk->b_datap == dbp);
+	ASSERT(msg_size != 0);
+	ASSERT(dbp->db_struioflag == 0);
+	ASSERT(dbp->db_struioun.cksum.flags == 0);
+
+	if ((msg_size & PAGEOFFSET) == 0) {
+		kmem_free(dbp->db_base, msg_size);
+	}
+
+	kmem_cache_free(mblk_cache, dbp->db_mblk);
+}
+
+static void
+bcache_dblk_destructor(void *buf, void *cdrarg)
+{
+	dblk_t *dbp = buf;
+	bcache_t *bcp = cdrarg;
+
+	kmem_cache_free(bcp->buffer_cache, dbp->db_base);
+
+	ASSERT(dbp->db_mblk->b_datap == dbp);
+	ASSERT(dbp->db_struioflag == 0);
+	ASSERT(dbp->db_struioun.cksum.flags == 0);
+
+	kmem_cache_free(mblk_cache, dbp->db_mblk);
+}
+
+#ifdef	_KERNEL
```

Comment:
Seems like this entire block could simply be removed?

Reply:
In most of these "fake" copies of things, I was trying hard to keep them
as close as possible to the originals (they're nearly identical).
My thought was that we might want to consider tweaking the
originals in some cases to avoid these copies.  If that doesn't
end up happening, then yes, a lot could be chopped out.